### ISSUE TITLE
Etcd advertises wrong client urls

### ISSUE THREAD

**Issue by mkudlej** [coreos/etcd#7588]
Etcd server advertises client urls as "http://localhost:2379" even if
it is set in systemd environment file to "http://10.34.108.73:2379".
Because of this clients(in my case etcdctl) which ask for list of
clients, cannot connect to Etcd because etcd is not listening on
localhost. As workaround for testing purpose I use `0.0.0.0` instead
of IP or hostname in configuration. Then Etcd listens on all
interfaces.

Etcdctl debug output:
```
$ etcdctl --debug --endpoints http://10.34.108.73:2379 ls /
start to sync cluster using endpoints(http://10.34.108.73:2379)
cURL Command: curl -X GET http://10.34.108.73:2379/v2/members
got endpoints(http://localhost:2379) after sync
Cluster-Endpoints: http://localhost:2379
cURL Command: curl -X GET http://localhost:2379/v2/keys/?quorum=false&recursive=false&sorted=false
Error:  client: etcd cluster is unavailable or misconfigured; error #0: dial tcp [::1]:2379: getsockopt: connection refused

error #0: dial tcp [::1]:2379: getsockopt: connection refused
```

I don't see this behaviour in previous version:
```
$ etcd --version
etcd Version: 3.0.15
Git SHA: fc00305
Go Version: go1.6.3
Go OS/Arch: linux/amd64
```

Client connection with workaround settings:
```
$ etcdctl --debug --endpoints http://10.34.108.73:2379 ls /
start to sync cluster using endpoints(http://10.34.108.73:2379)
cURL Command: curl -X GET http://10.34.108.73:2379/v2/members
got endpoints(http://0.0.0.0:2379) after sync
Cluster-Endpoints: http://0.0.0.0:2379
cURL Command: curl -X GET http://0.0.0.0:2379/v2/keys/?quorum=false&recursive=false&sorted=false
/nodes
/queue
/Messages
/indexes
/monitoring
/networks
```

but this doens't work from remote machine of course.

Additional details:
```
$ etcd --version
etcd Version: 3.1.0
Git SHA: 8ba2897
Go Version: go1.7.4
Go OS/Arch: linux/amd64
```
```
$ systemctl cat etcd
# /usr/lib/systemd/system/etcd.service
[Unit]
Description=Etcd Server
After=network.target
After=network-online.target
Wants=network-online.target

[Service]
Type=notify
WorkingDirectory=/var/lib/etcd/
EnvironmentFile=-/etc/etcd/etcd.conf
User=etcd
# set GOMAXPROCS to number of processors
ExecStart=/bin/bash -c "GOMAXPROCS=$(nproc) /usr/bin/etcd --name=\"${ETCD_NAME}\" --data-dir=\"${ETCD_DATA_DIR}\" --listen-client-urls=\"${ETCD_LISTEN_CLIENT_URLS}\""
Restart=on-failure
LimitNOFILE=65536

[Install]
WantedBy=multi-user.target
```
```
#restart etcd service
$ systemctl restart etcd
$ journalctl -u etcd
# where _hostname_ is private hostname for 10.34.108.73
# as you can see here even if advertise client URLs is set
# to ETCD_LISTEN_CLIENT_URLS="http://10.34.108.73:2379"
# during service start it is set to "advertise client URLs =
# http://localhost:2379"
Mar 23 20:18:11 _hostname_ systemd[1]: Starting Etcd Server...
Mar 23 20:18:11 _hostname_ etcd[24998]: recognized and used environment variable ETCD_ADVERTISE_CLIENT_URLS=http://10.34.108.73:2379
Mar 23 20:18:11 _hostname_ etcd[24998]: recognized and used environment variable ETCD_INITIAL_ADVERTISE_PEER_URLS=http://10.34.108.73:2380
Mar 23 20:18:11 _hostname_ etcd[24998]: recognized and used environment variable ETCD_INITIAL_CLUSTER=default=http://10.34.108.73:2380
Mar 23 20:18:11 _hostname_ etcd[24998]: recognized and used environment variable ETCD_LISTEN_PEER_URLS=http://10.34.108.73:2380
Mar 23 20:18:11 _hostname_ etcd[24998]: recognized environment variable ETCD_NAME, but unused: shadowed by corresponding flag 
Mar 23 20:18:11 _hostname_ etcd[24998]: recognized environment variable ETCD_DATA_DIR, but unused: shadowed by corresponding flag 
Mar 23 20:18:11 _hostname_ etcd[24998]: recognized environment variable ETCD_LISTEN_CLIENT_URLS, but unused: shadowed by corresponding flag 
Mar 23 20:18:11 _hostname_ etcd[24998]: etcd Version: 3.1.0
Mar 23 20:18:11 _hostname_ etcd[24998]: Git SHA: 8ba2897
Mar 23 20:18:11 _hostname_ etcd[24998]: Go Version: go1.7.4
Mar 23 20:18:11 _hostname_ etcd[24998]: Go OS/Arch: linux/amd64
Mar 23 20:18:11 _hostname_ etcd[24998]: setting maximum number of CPUs to 2, total number of available CPUs is 2
Mar 23 20:18:11 _hostname_ etcd[24998]: the server is already initialized as member before, starting as etcd member...
Mar 23 20:18:11 _hostname_ etcd[24998]: advertising using detected default host "10.34.108.73"
Mar 23 20:18:11 _hostname_ etcd[24998]: listening for peers on http://10.34.108.73:2380
Mar 23 20:18:11 _hostname_ etcd[24998]: listening for client requests on 10.34.108.73:2379
Mar 23 20:18:11 _hostname_ etcd[24998]: recovered store from snapshot at index 15342988
Mar 23 20:18:11 _hostname_ etcd[24998]: name = default
Mar 23 20:18:11 _hostname_ etcd[24998]: data dir = /var/lib/etcd/default.etcd
Mar 23 20:18:11 _hostname_ etcd[24998]: member dir = /var/lib/etcd/default.etcd/member
Mar 23 20:18:11 _hostname_ etcd[24998]: heartbeat = 100ms
Mar 23 20:18:11 _hostname_ etcd[24998]: election = 1000ms
Mar 23 20:18:11 _hostname_ etcd[24998]: snapshot count = 10000
Mar 23 20:18:11 _hostname_ etcd[24998]: advertise client URLs = http://localhost:2379
Mar 23 20:18:12 _hostname_ etcd[24998]: restarting member 1b427d1a51d974c9 in cluster 3b0ca9fc906e999a at commit index 15351716
Mar 23 20:18:12 _hostname_ etcd[24998]: 1b427d1a51d974c9 became follower at term 9
Mar 23 20:18:12 _hostname_ etcd[24998]: newRaft 1b427d1a51d974c9 [peers: [1b427d1a51d974c9], term: 9, commit: 15351716, applied: 15342988, lastindex: 15
Mar 23 20:18:12 _hostname_ etcd[24998]: enabled capabilities for version 3.1
Mar 23 20:18:12 _hostname_ etcd[24998]: added member 1b427d1a51d974c9 [http://10.34.108.73:2380] to cluster 3b0ca9fc906e999a from store
Mar 23 20:18:12 _hostname_ etcd[24998]: set the cluster version to 3.1 from store
Mar 23 20:18:12 _hostname_ etcd[24998]: starting server... [version: 3.1.0, cluster version: 3.1]
Mar 23 20:18:12 _hostname_ etcd[24998]: 1b427d1a51d974c9 is starting a new election at term 9
Mar 23 20:18:12 _hostname_ etcd[24998]: 1b427d1a51d974c9 became candidate at term 10
Mar 23 20:18:12 _hostname_ etcd[24998]: 1b427d1a51d974c9 received MsgVoteResp from 1b427d1a51d974c9 at term 10
Mar 23 20:18:12 _hostname_ etcd[24998]: 1b427d1a51d974c9 became leader at term 10
Mar 23 20:18:12 _hostname_ etcd[24998]: raft.node: 1b427d1a51d974c9 elected leader 1b427d1a51d974c9 at term 10
Mar 23 20:18:13 _hostname_ etcd[24998]: published {Name:default ClientURLs:[http://localhost:2379]} to cluster 3b0ca9fc906e999a
Mar 23 20:18:13 _hostname_ etcd[24998]: ready to serve client requests
Mar 23 20:18:13 _hostname_ etcd[24998]: serving insecure client requests on 10.34.108.73:2379, this is strongly discouraged!
Mar 23 20:18:13 _hostname_ systemd[1]: Started Etcd Server.
```
```
$ cat /etc/etcd/etcd.conf
# [member]
ETCD_NAME=default
ETCD_DATA_DIR="/var/lib/etcd/default.etcd"
#ETCD_WAL_DIR=""
#ETCD_SNAPSHOT_COUNT="10000"
#ETCD_HEARTBEAT_INTERVAL="100"
#ETCD_ELECTION_TIMEOUT="1000"
ETCD_LISTEN_PEER_URLS="http://10.34.108.73:2380"
ETCD_LISTEN_CLIENT_URLS="http://10.34.108.73:2379"
#ETCD_MAX_SNAPSHOTS="5"
#ETCD_MAX_WALS="5"
#ETCD_CORS=""
#
#[cluster]
ETCD_INITIAL_ADVERTISE_PEER_URLS="http://10.34.108.73:2380"
# if you use different ETCD_NAME (e.g. test), set ETCD_INITIAL_CLUSTER value for this name, i.e. "test=http://..."
ETCD_INITIAL_CLUSTER="default=http://10.34.108.73:2380"
#ETCD_INITIAL_CLUSTER_STATE="new"
#ETCD_INITIAL_CLUSTER_TOKEN="etcd-cluster"
ETCD_ADVERTISE_CLIENT_URLS="http://10.34.108.73:2379"
... all other options are commented
```

**Comment by heyitsanthony** [comment]
@mkudlej this should be fixed via 059dc91d4cee5e88991bf082e624aa3f33eced1f. Try 3.1.3?

**Comment by mkudlej** [comment]
@heyitsanthony You are right, it is fixed in:
```
$ etcd --version
etcd Version: 3.1.3
Git SHA: 21fdcc6
Go Version: go1.7.4
Go OS/Arch: linux/amd64
```
